File: app/src/main/java/de/westnordost/streetcomplete/data/osm/OsmQuestGiver.java
package de.westnordost.streetcomplete.data.osm;

import android.text.TextUtils;
import android.util.Log;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import javax.inject.Inject;
import javax.inject.Provider;

import de.westnordost.osmapi.map.data.BoundingBox;
import de.westnordost.osmapi.map.data.Element;
import de.westnordost.osmapi.map.data.LatLon;
import de.westnordost.streetcomplete.data.QuestStatus;
import de.westnordost.streetcomplete.data.QuestType;
import de.westnordost.streetcomplete.data.osm.persist.ElementGeometryDao;
import de.westnordost.streetcomplete.data.osm.persist.OsmQuestDao;
import de.westnordost.streetcomplete.data.osmnotes.OsmNoteQuestDao;

/** Manages creating new quests and removing quests that are no longer applicable for an OSM
 *  element locally */
public class OsmQuestGiver
{
	private static final String TAG = "OsmQuestGiver";

	private final OsmNoteQuestDao osmNoteQuestDb;
	private final OsmQuestDao questDB;
	private final ElementGeometryDao elementGeometryDB;
	private final Provider<List<QuestType>> questTypesProvider;

	@Inject public OsmQuestGiver(
			OsmNoteQuestDao osmNoteQuestDb, OsmQuestDao questDB,
			ElementGeometryDao elementGeometryDB, Provider<List<QuestType>> questTypesProvider)
	{
		this.osmNoteQuestDb = osmNoteQuestDb;
		this.questDB = questDB;
		this.elementGeometryDB = elementGeometryDB;
		this.questTypesProvider = questTypesProvider;
	}

	public static class QuestUpdates
	{
		public List<OsmQuest> createdQuests = new ArrayList<>();
		public List<Long> removedQuestIds = new ArrayList<>();
	}

	public QuestUpdates updateQuests(Element element)
	{
		ElementGeometry geometry = elementGeometryDB.get(element.getType(), element.getId());
		boolean hasNote = hasNoteAt(geometry.center);

		QuestUpdates result = new QuestUpdates();

		Map<QuestType, OsmQuest> currentQuests = getCurrentQuests(element);
		List<String> createdQuestsLog = new ArrayList<>();
		List<String> removedQuestsLog = new ArrayList<>();

		for(QuestType questType : questTypesProvider.get())
		{
			if(!(questType instanceof OsmElementQuestType)) continue;
			OsmElementQuestType osmQuestType = (OsmElementQuestType)questType;

			Boolean appliesToElement = osmQuestType.isApplicableTo(element);
			if(appliesToElement == null) continue;

			boolean hasQuest = currentQuests.containsKey(osmQuestType);
			if(appliesToElement && !hasQuest && !hasNote)
			{
				OsmQuest quest = new OsmQuest(osmQuestType, element.getType(), element.getId(), geometry);
				result.createdQuests.add(quest);
				createdQuestsLog.add(osmQuestType.getClass().getSimpleName());
			}
			if(!appliesToElement && hasQuest)
			{
				OsmQuest quest = currentQuests.get(osmQuestType);
				// only remove "fresh" unanswered quests because answered/closed quests by definition
				// do not apply to the element anymore. E.g. after adding the name to the street,
				// there shan't be any AddRoadName quest for that street anymore
				if(quest.getStatus() == QuestStatus.NEW)
				{
					result.removedQuestIds.add(quest.getId());
					removedQuestsLog.add(osmQuestType.getClass().getSimpleName());
				}
			}
		}

		if(!result.createdQuests.isEmpty())
		{
			// Before new quests are unlocked, all reverted quests need to be removed for
			// this element so that they can be created anew as the case may be
			questDB.deleteAllReverted(element.getType(), element.getId());

			questDB.addAll(result.createdQuests);

			Log.d(TAG, "Created new quests for " + element.getType().name() + "#" + element.getId() + ": " +
				TextUtils.join(", ", createdQuestsLog)
			);
		}
		if(!result.removedQuestIds.isEmpty())
		{
			questDB.deleteAll(result.removedQuestIds);

			Log.d(TAG, "Removed quests no longer applicable for " + element.getType().name() + "#" + element.getId() + ": " +
				TextUtils.join(", ", removedQuestsLog)
			);
		}

		return result;
	}

	private boolean hasNoteAt(LatLon pos)
	{
		BoundingBox bbox = new BoundingBox(pos, pos);
		return !osmNoteQuestDb.getAllPositions(bbox).isEmpty();
	}

	private Map<QuestType, OsmQuest> getCurrentQuests(Element element)
	{
		List<OsmQuest> quests = questDB.getAll(null, null, null, element.getType(), element.getId());
		Map<QuestType, OsmQuest> result = new HashMap<>(quests.size());
		for (OsmQuest quest : quests)
		{
			if(quest.getStatus() == QuestStatus.REVERT) continue;
			result.put(quest.getType(), quest);
		}
		return result;
	}
}

File: app/src/main/java/de/westnordost/streetcomplete/data/osm/OsmQuestGiver.kt
package de.westnordost.streetcomplete.data.osm

import android.util.Log
import de.westnordost.countryboundaries.CountryBoundaries
import de.westnordost.countryboundaries.isInAny


import javax.inject.Inject

import de.westnordost.osmapi.map.data.Element
import de.westnordost.osmapi.map.data.LatLon
import de.westnordost.streetcomplete.data.QuestStatus
import de.westnordost.streetcomplete.data.QuestType
import de.westnordost.streetcomplete.data.osm.persist.ElementGeometryDao
import de.westnordost.streetcomplete.data.osm.persist.OsmQuestDao
import de.westnordost.streetcomplete.data.osmnotes.OsmNoteQuestDao
import de.westnordost.streetcomplete.data.visiblequests.OrderedVisibleQuestTypesProvider
import de.westnordost.streetcomplete.util.enclosingBoundingBox
import java.util.concurrent.FutureTask

/** Manages creating new quests and removing quests that are no longer applicable for an OSM
 * element locally  */
class OsmQuestGiver @Inject constructor(
    private val osmNoteQuestDb: OsmNoteQuestDao,
    private val questDB: OsmQuestDao,
    private val elementGeometryDB: ElementGeometryDao,
    private val questTypesProvider: OrderedVisibleQuestTypesProvider,
    private val countryBoundariesFuture: FutureTask<CountryBoundaries>
) {

    data class QuestUpdates(val createdQuests: List<OsmQuest>, val removedQuestIds: List<Long>)

    fun updateQuests(element: Element): QuestUpdates {
        val createdQuests: MutableList<OsmQuest> = ArrayList()
        val removedQuestIds: MutableList<Long> = ArrayList()

        val geometry = elementGeometryDB.get(element.type, element.id) ?: return QuestUpdates(
            emptyList(), emptyList()
        )

        val hasNote = hasNoteAt(geometry.center)

        val currentQuests = getCurrentQuests(element)
        val createdQuestsLog = ArrayList<String>()
        val removedQuestsLog = ArrayList<String>()

        for (questType in questTypesProvider.get()) {
            if (questType !is OsmElementQuestType<*>) continue

            val appliesToElement = questType.isApplicableTo(element) ?: continue
            val countries = questType.enabledInCountries
            val isEnabledForCountry = countryBoundariesFuture.get().isInAny(geometry.center, countries)

            val hasQuest = currentQuests.containsKey(questType)
            if (appliesToElement && !hasQuest && !hasNote && isEnabledForCountry) {
                val quest = OsmQuest(questType, element.type, element.id, geometry)
                createdQuests.add(quest)
                createdQuestsLog.add(questType.javaClass.simpleName)
            }
            if (!appliesToElement && hasQuest) {
                val quest = currentQuests.getValue(questType)
                // only remove "fresh" unanswered quests because answered/closed quests by definition
                // do not apply to the element anymore. E.g. after adding the name to the street,
                // there shan't be any AddRoadName quest for that street anymore
                if (quest.status == QuestStatus.NEW) {
                    removedQuestIds.add(quest.id!!)
                    removedQuestsLog.add(questType.javaClass.simpleName)
                }
            }
        }

        if (createdQuests.isNotEmpty()) {
            // Before new quests are unlocked, all reverted quests need to be removed for
            // this element so that they can be created anew as the case may be
            questDB.deleteAll(
                statusIn = listOf(QuestStatus.REVERT),
                element = ElementKey(element.type, element.id)
            )
            questDB.addAll(createdQuests)
            Log.d(TAG, "Created new quests for ${element.type.name}#${element.id}: ${createdQuestsLog.joinToString()}")
        }
        if (removedQuestIds.isNotEmpty()) {
            questDB.deleteAllIds(removedQuestIds)
            Log.d(TAG, "Removed quests no longer applicable for ${element.type.name}#${element.id}: ${removedQuestsLog.joinToString()}")
        }

        return QuestUpdates(createdQuests, removedQuestIds)
    }

    fun deleteQuests(elementType: Element.Type, elementId: Long): List<Long> {
        val ids = questDB.getAllIds(element = ElementKey(elementType, elementId))
        questDB.deleteAllIds(ids)

        Log.d(TAG, "Removed all quests for deleted element " + elementType.name + "#" + elementId)
        return ids
    }

    private fun hasNoteAt(pos: LatLon): Boolean {
        // note about one meter around the center of an element still count as at this point as to
        // deal with imprecision of the center calculation of geometry (see #1089)
        val bbox = pos.enclosingBoundingBox(1.0)
        return osmNoteQuestDb.getAllPositions(bbox).isNotEmpty()
    }

    private fun getCurrentQuests(element: Element): Map<QuestType<*>, OsmQuest> {
        val quests = questDB.getAll(element = ElementKey(element.type, element.id))
        val result = HashMap<QuestType<*>, OsmQuest>(quests.size)
        for (quest in quests) {
            if (quest.status == QuestStatus.REVERT) continue
            result[quest.type] = quest
        }
        return result
    }

    companion object {
        private const val TAG = "OsmQuestGiver"
    }
}
